### PR TITLE
promotion: migrate to targets, tolerate mixed config

### PR DESCRIPTION
**Pull request by stevekuznetsov**
This commit updates the determinizer to migrate ci-operator configs to use the new promotion.to[] stanza instead of the previous inline configuration. For now, we will continue to honor the older configuration, a mixed configuration or only the new one.

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Looking into integration

**Comment by stevekuznetsov** [comment]
@bbguimaraes the `test/integration: determinize ci-operator configs` did not run integration with `UPDATE=true,` just the determinizer

**Comment by stevekuznetsov** [comment]
@bbguimaraes the last commit ends up because `repo-init` does not write determinized prow config - would appreciate your help here, I have limited bandwidth to go digging into the repo right now :(

File: cmd/promoted-image-governor/main.go [hunk 1]
```diff
@@ -441,7 +441,11 @@ func main() {
 			if _, ok := opts.explains[isTagRef]; ok {
 				opts.explains[isTagRef] = cfg.Metadata.AsString()
 			}
-			if cfg.PromotionConfiguration.TagByCommit {
+			var taggedByCommit bool
+			for _, target := range api.PromotionTargets(cfg.PromotionConfiguration) {
+				taggedByCommit = taggedByCommit || target.TagByCommit
+			}
```
Comment (bbguimaraes) [inline cmd/promoted-image-governor/main.go:447]: This checks if _any_ tag is created by commit, not just `isTagRef`.

File: pkg/image-graph-generator/operator.go [hunk 2]
```diff
@@ -80,12 +80,14 @@ func (o *Operator) callback(c *api.ReleaseBuildConfiguration, i *config.Info) er
...
-
-	for _, image := range c.Images {
-		if !excludedImages.Has(string(image.To)) {
-			if err := o.UpdateImage(image, c, branchID); err != nil {
-				errs = append(errs, err)
+	for _, target := range api.PromotionTargets(c.PromotionConfiguration) {
+		excludedImages := sets.New[string](target.ExcludedImages...)
+
+		for _, image := range c.Images {
+			if !excludedImages.Has(string(image.To)) {
```
Comment (bbguimaraes) [inline pkg/image-graph-generator/operator.go:87]: This has to take into account the new meaning of `*`.

Comment (bbguimaraes) [reply pkg/image-graph-generator/operator.go:87]: The existing code also doesn't seem to handle `additional_images`.

File: cmd/config-brancher/main.go [hunk 3]
```diff
@@ -107,7 +107,7 @@ func generateBranchedConfigs(currentRelease, bumpRelease string, futureReleases
 	devRelease := currentRelease
 	if bumpRelease != "" && promotion.IsBumpable(input.Info.Branch, currentRelease) {
 		devRelease = bumpRelease
-		updateRelease(&currentConfig, bumpRelease)
+		updateRelease(&currentConfig, currentRelease, bumpRelease)
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply cmd/config-brancher/main.go:110]: Sorry, I don't quite understand the comment - isn't conditionally updating the specific `PromotionConfiguration.Targets` entry what my code does in `updateRelease` now?

File: cmd/promoted-image-governor/main.go (hunk 1 above)
Comment (stevekuznetsov) [reply cmd/promoted-image-governor/main.go:447]: Yes, I'm happy to make an issue for someone to make that better in the future but I am at the boundary of how much I can refactor here. The regex created will be a superset of the regex that's needed, and if for some reason images tagged by PR from another `Targets` stanza existed, they would not get cleaned up. The refactor here would require that `release.PromotedTags` return something other than a flat `isTagRef` list.

File: pkg/image-graph-generator/operator.go (hunk 2 above)
Comment (stevekuznetsov) [reply pkg/image-graph-generator/operator.go:87]: I have yet to code in the `*` bit - and as for `additional_images` here, I'm not sure exactly how to proceed, all I did in this diff was move the code into a loop over targets; so the previous code also did not handle `additional_images` - should it have?

**Comment by stevekuznetsov** [comment]
/test e2e

Some ImagePullBackOff etc

File: cmd/promoted-image-governor/main.go (hunk 1 above)
Comment (bbguimaraes) [reply cmd/promoted-image-governor/main.go:447]: We can iterate over `PromotionTargets`, which is where all the information we need is.

There would still be some ways to create an unfortunate configuration which would not be processed correctly, but that's already the case for the current implementation (before `Targets`).

File: cmd/config-brancher/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply cmd/config-brancher/main.go:110]: This is now a possible configuration:

```yaml
promotion:
  namespace: x
  name: y
  to:
  - namespace: ocp
    name: "4.14"
```

We cannot just modify the (top-level) `name` unconditionally anymore.

File: pkg/image-graph-generator/operator.go (hunk 2 above)
Comment (bbguimaraes) [reply pkg/image-graph-generator/operator.go:87]: I was hoping we could just use the same code that's used in other places to determine the images to be promoted and get rid of this partial (re-)implementation.

File: cmd/promoted-image-governor/main.go (hunk 1 above)
Comment (stevekuznetsov) [reply cmd/promoted-image-governor/main.go:447]: (Not impossible, just asking for this to be a follow-up that someone on DPTP could handle?)

File: pkg/image-graph-generator/operator.go (hunk 2 above)
Comment (stevekuznetsov) [reply pkg/image-graph-generator/operator.go:87]: (Not impossible, just asking for this to be a follow-up that someone on DPTP could handle?)

File: cmd/config-brancher/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply cmd/config-brancher/main.go:110]: If we're not going to handle both, then this is fine.

File: cmd/promoted-image-governor/main.go (hunk 1 above)
Comment (bbguimaraes) [reply cmd/promoted-image-governor/main.go:447]: I see `PromotedTags` is quite messy already.  At least leave a comment here, then.

File: pkg/image-graph-generator/operator.go (hunk 2 above)
Comment (bbguimaraes) [reply pkg/image-graph-generator/operator.go:87]: Leave a comment here too.  There's no reason for this code duplication.

**Comment by danilo-gemoli** [comment]
/retest
if it doesn't make it next run I'm going to `/override` as that's not an error

**Comment by stevekuznetsov** [comment]
```
   * could not run steps: step src failed: error occurred handling build src: could not delete build src: the server is currently unable to handle the request (get builds.build.openshift.io src) 
```

/test e2e

**Comment by stevekuznetsov** [comment]
```
 time="2023-09-19T11:16:33Z" level=error msg="failed to execute step" error="couldn't find SecretItem with item_name: build_farm name: sa.$(service_account).$(cluster).config containing cluster: build01" 
```

What is that in breaking changes job?

**Comment by danilo-gemoli** [comment]
The test is failing due to https://github.com/openshift/release/pull/43448; that PR was opened to cope with `build01` outage.
/override ci/prow/breaking-changes
/lgtm

**Pull request merged**